File: backend_inventra/models/mutasi.js
const mongoose = require('mongoose');

const mutasiStokSchema = new mongoose.Schema({
    barang_id: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Barang',
        required: true
    },
    tipe: {
        type: String,
        enum: ['masuk', 'keluar', 'opname', 'retur'],
        required: true
    },
    jumlah: {
        type: Number,
        required: true,
        min: 1
    },
    keterangan: {
        type: String,
        required: true
    },
    stok_sebelum: {
        type: Number,
        required: true
    },
    stok_sesudah: {
        type: Number,
        required: true
    },
    created_by: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    tanggal_mutasi: {
        type: Date,
        default: Date.now
    },
    created_at: {
        type: Date,
        default: Date.now
    }
}, {
    collection: 'mutasi_stok'
});

// Index untuk query cepat
mutasiStokSchema.index({ barang_id: 1, tanggal_mutasi: -1 });
mutasiStokSchema.index({ created_by: 1 });

module.exports = mongoose.model('MutasiStok', mutasiStokSchema);

File: flutter_inventra/lib/models/mutasi.dart
class Mutasi {
  final String id;
  final String barangId;
  final String kodeBarang;
  final String namaBarang;
  final String tipe;
  final int jumlah;
  final int stokSebelum;
  final int stokSesudah;
  final String keterangan;
  final String tanggalMutasi;
  final String createdBy;

  const Mutasi({
    required this.id,
    required this.barangId,
    required this.kodeBarang,
    required this.namaBarang,
    required this.tipe,
    required this.jumlah,
    required this.stokSebelum,
    required this.stokSesudah,
    required this.keterangan,
    required this.tanggalMutasi,
    required this.createdBy,
  });

  factory Mutasi.fromJson(Map<String, dynamic> json) {
    final barang =
        json['barang_id'] is Map<String, dynamic>
            ? json['barang_id'] as Map<String, dynamic>
            : <String, dynamic>{};
    final createdBy =
        json['created_by'] is Map<String, dynamic>
            ? json['created_by'] as Map<String, dynamic>
            : <String, dynamic>{};

    return Mutasi(
      id: (json['_id'] ?? '').toString(),
      barangId: (barang['_id'] ?? json['barang_id'] ?? '').toString(),
      kodeBarang: (barang['kode_barang'] ?? '').toString(),
      namaBarang: (barang['nama_barang'] ?? '').toString(),
      tipe: (json['tipe'] ?? '').toString(),
      jumlah: (json['jumlah'] as num? ?? 0).toInt(),
      stokSebelum: (json['stok_sebelum'] as num? ?? 0).toInt(),
      stokSesudah: (json['stok_sesudah'] as num? ?? 0).toInt(),
      keterangan: (json['keterangan'] ?? '').toString(),
      tanggalMutasi: (json['tanggal_mutasi'] ?? '').toString(),
      createdBy: (createdBy['nama'] ?? '').toString(),
    );
  }
}
